# Implementing PCA in Python with Scikit-Learn
https://stackabuse.com/implementing-pca-in-python-with-scikit-learn/

With the availability of high performance CPUs and GPUs, it is pretty much possible to solve every regression, classification, clustering and other related problems using machine learning and deep learning models. However, there are still various factors that cause performance bottlenecks while developing such models. Large number of features in the dataset is one of the factors that affect both the training time as well as accuracy of machine learning models. You have different options to deal with huge number of features in a dataset.

Try to train the models on original number of features, which take days or weeks if the number of features is too high.
Reduce the number of variables by merging correlated variables.
Extract the most important features from the dataset that are responsible for maximum variance in the output. Different statistical techniques are used for this purpose e.g. linear discriminant analysis, factor analysis, and principal component analysis.
In this article, we will see how principal component analysis can be implemented using Python's Scikit-Learn library.

# Principal Component Analysis

Principal component analysis, or PCA, is a statistical technique to convert high dimensional data to low dimensional data by selecting the most important features that capture maximum information about the dataset. The features are selected on the basis of variance that they cause in the output. The feature that causes highest variance is the first principal component. The feature that is responsible for second highest variance is considered the second principal component, and so on. It is important to mention that principal components do not have any correlation with each other.

# Advantages of PCA

There are two main advantages of dimensionality reduction with PCA.

The training time of the algorithms reduces significantly with less number of features.
It is not always possible to analyze data in high dimensions. For instance if there are 100 features in a dataset. Total number of scatter plots required to visualize the data would be 100(100-1)2 = 4950. Practically it is not possible to analyze data this way.

# Normalization of Features

It is imperative to mention that a feature set must be normalized before applying PCA. For instance if a feature set has data expressed in units of Kilograms, Light years, or Millions, the variance scale is huge in the training set. If PCA is applied on such a feature set, the resultant loadings for features with high variance will also be large. Hence, principal components will be biased towards features with high variance, leading to false results.

Finally, the last point to remember before we start coding is that PCA is a statistical technique and can only be applied to numeric data. Therefore, categorical features are required to be converted into numerical features before PCA can be applied.

# Implementing PCA with Scikit-Learn
In this section we will implement PCA with the help of Python's Scikit-Learn library. We will follow the classic machine learning pipeline where we will first import libraries and dataset, perform exploratory data analysis and preprocessing, and finally train our models, make predictions and evaluate accuracies. The only additional step will be to perform PCA to find out optimal number of features before we train our models. These steps have been implemented as follows:

Importing Dataset
The dataset we are going to use in this article is the famous Iris data set. Some additional information about the Iris dataset is available at:

https://archive.ics.uci.edu/ml/datasets/iris

The dataset consists of 150 records of Iris plant with four features: 'sepal-length', 'sepal-width', 'petal-length', and 'petal-width'. All of the features are numeric. The records have been classified into one of the three classes i.e. 'Iris-setosa', 'Iris-versicolor', or 'Iris-verginica'.

Execute the following script to download the dataset using pandas:

In [0]:

import numpy as np
import pandas as pd

In [0]:
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"
names = ['sepal-length', 'sepal-width', 'petal-length', 'petal-width', 'Class']
dataset = pd.read_csv(url, names=names)
dataset.head()

sepal-length  sepal-width  petal-length  petal-width        Class
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa
2           4.7          3.2           1.3          0.2  Iris-setosa
3           4.6          3.1           1.5          0.2  Iris-setosa
4           5.0          3.6           1.4          0.2  Iris-setosa

# Preprocessing
The first preprocessing step is to divide the dataset into a feature set and corresponding labels. The following script performs this task:

In [0]:
X = dataset.drop('Class',1)
y = dataset['Class']

The script above stores the feature sets into the X variable and the series of corresponding labels in to the y variable.

The next preprocessing step is to divide data into training and test sets. Execute the following script to do so:

In [0]:
# Splitting the dataset into the Training set and Test set
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

# standardization

As mentioned earlier, PCA performs best with a normalized feature set. We will perform standard scalar normalization to normalize our feature set. To do this, execute the following code:

In [0]:
from sklearn.preprocessing import StandardScaler

sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

# Applying PCA
It is only a matter of three lines of code to perform PCA using Python's Scikit-Learn library. The PCA class is used for this purpose. PCA depends only upon the feature set and not the label data. Therefore, PCA can be considered as an unsupervised machine learning technique.

Performing PCA using Scikit-Learn is a two-step process:

<br>1- Initialize the PCA class by passing the number of components to the constructor.</br>
<br>2- Call the fit and then transform methods by passing the feature set to these methods. The transform method returns the
specified number of principal components.</br> 
Take a look at the following code:

In [0]:
from sklearn.decomposition import PCA

pca = PCA()
X_train = pca.fit_transform(X_train)
X_test = pca.transform(X_test)

In the code above, we create a PCA object named pca. We did not specify the number of components in the constructor. Hence, all four of the features in the feature set will be returned for both the training and test sets.

The PCA class contains explained_variance_ratio_ which returns the variance caused by each of the principal components. Execute the following line of code to find the "explained variance ratio".

In [0]:
explained_variance = pca.explained_variance_ratio_
explained_variance

array([ 0.72226528,  0.23974795,  0.03338117,  0.0046056 ])

The explained_variance variable is now a float type array which contains variance ratios for each principal component. The values for the explained_variance variable looks like this:[ 0.72226528,  0.23974795,  0.03338117,  0.0046056 ]

It can be seen that first principal component is responsible for 72.22% variance. Similarly, the second principal component causes 23.9% variance in the dataset. Collectively we can say that (72.22 + 23.9) 96.21% percent of the classification information contained in the feature set is captured by the first two principal components.

<h1>Let's first try to use 1 principal component to train our algorithm. To do so, execute the following code: this means other 3 features will project on the first one.</h1>


In [0]:
from sklearn.decomposition import PCA

pca = PCA(n_components=1)
X_train = pca.fit_transform(X_train)
X_test = pca.transform(X_test)

 The rest of the process is straight forward.

# Training and Making Predictions
In this case we'll use random forest classification for making the predictions.

In [0]:
from sklearn.ensemble import RandomForestClassifier

classifier = RandomForestClassifier(max_depth=2, random_state=0)
classifier.fit(X_train, y_train)

# Predicting the Test set results
y_pred = classifier.predict(X_test)

In [0]:
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score

cm = confusion_matrix(y_test, y_pred)
print(cm)
print(accuracy_score(y_test, y_pred))

[[11  0  0]
 [ 0 12  1]
 [ 0  1  5]]
0.933333333333


It can be seen from the output that with only one feature, the random forest algorithm is able to correctly predict 28 out of 30 instances, resulting in 93.33% accuracy.

# Results with 2 components that means project other fetures in two vector. so it would be 2dimension.
Now let's try to evaluate classification performance of the random forest algorithm with 2 principal components. Update this piece of code:

In [0]:
import numpy as np
import pandas as pd
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"
names = ['sepal-length', 'sepal-width', 'petal-length', 'petal-width', 'Class']
dataset = pd.read_csv(url, names=names)
dataset.head()
X1 = dataset.drop('Class',1)
y1 = dataset['Class']
X1

sepal-length  sepal-width  petal-length  petal-width
0             5.1          3.5           1.4          0.2
1             4.9          3.0           1.4          0.2
2             4.7          3.2           1.3          0.2
3             4.6          3.1           1.5          0.2
4             5.0          3.6           1.4          0.2
5             5.4          3.9           1.7          0.4
6             4.6          3.4           1.4          0.3
7             5.0          3.4           1.5          0.2
8             4.4          2.9           1.4          0.2
9             4.9          3.1           1.5          0.1
10            5.4          3.7           1.5          0.2
11            4.8          3.4           1.6          0.2
12            4.8          3.0           1.4          0.1
13            4.3          3.0           1.1          0.1
14            5.8          4.0           1.2          0.2
15            5.7          4.4           1.5          0.4
16            5.4          3.9           1.3          0.4
17            5.1          3.5           1.4          0.3
18            5.7          3.8           1.7          0.3
19            5.1          3.8           1.5          0.3
20            5.4          3.4           1.7          0.2
21            5.1          3.7           1.5          0.4
22            4.6          3.6           1.0          0.2
23            5.1          3.3           1.7          0.5
24            4.8          3.4           1.9          0.2
25            5.0          3.0           1.6          0.2
26            5.0          3.4           1.6          0.4
27            5.2          3.5           1.5          0.2
28            5.2          3.4           1.4          0.2
29            4.7          3.2           1.6          0.2
..            ...          ...           ...          ...
120           6.9          3.2           5.7          2.3
121           5.6          2.8           4.9          2.0
122           7.7          2.8           6.7          2.0
123           6.3          2.7           4.9          1.8
124           6.7          3.3           5.7          2.1
125           7.2          3.2           6.0          1.8
126           6.2          2.8           4.8          1.8
127           6.1          3.0           4.9          1.8
128           6.4          2.8           5.6          2.1
129           7.2          3.0           5.8          1.6
130           7.4          2.8           6.1          1.9
131           7.9          3.8           6.4          2.0
132           6.4          2.8           5.6          2.2
133           6.3          2.8           5.1          1.5
134           6.1          2.6           5.6          1.4
135           7.7          3.0           6.1          2.3
136           6.3          3.4           5.6          2.4
137           6.4          3.1           5.5          1.8
138           6.0          3.0           4.8          1.8
139           6.9          3.1           5.4          2.1
140           6.7          3.1           5.6          2.4
141           6.9          3.1           5.1          2.3
142           5.8          2.7           5.1          1.9
143           6.8          3.2           5.9          2.3
144           6.7          3.3           5.7          2.5
145           6.7          3.0           5.2          2.3
146           6.3          2.5           5.0          1.9
147           6.5          3.0           5.2          2.0
148           6.2          3.4           5.4          2.3
149           5.9          3.0           5.1          1.8

[150 rows x 4 columns]

In [0]:
from sklearn.model_selection import train_test_split

X1_train, X1_test, y1_train, y1_test = train_test_split(X1, y1, test_size=0.2, random_state=0)


# in this step we will scale all four features in train and test. 

In [0]:
from sklearn.preprocessing import StandardScaler

sc1 = StandardScaler()
X1_train = sc1.fit_transform(X1_train)
X1_test = sc1.transform(X1_test)
X1_train

array([[ 0.61303014,  0.10850105,  0.94751783,  0.73603967],
       [-0.56776627, -0.12400121,  0.38491447,  0.34808318],
       [-0.80392556,  1.03851009, -1.30289562, -1.3330616 ],
       [ 0.25879121, -0.12400121,  0.60995581,  0.73603967],
       [ 0.61303014, -0.58900572,  1.00377816,  1.25331499],
       [-0.80392556, -0.82150798,  0.04735245,  0.21876435],
       [-0.21352735,  1.73601687, -1.19037495, -1.20374277],
       [ 0.14071157, -0.82150798,  0.72247648,  0.47740201],
       [ 0.02263193, -0.12400121,  0.21613346,  0.34808318],
       [-0.09544771, -1.05401024,  0.10361279, -0.03987331],
       [ 1.0853487 , -0.12400121,  0.94751783,  1.12399616],
       [-1.39432376,  0.34100331, -1.41541629, -1.3330616 ],
       [ 1.20342834,  0.10850105,  0.72247648,  1.38263382],
       [-1.04008484,  1.03851009, -1.24663528, -0.81578628],
       [-0.56776627,  1.50351461, -1.30289562, -1.3330616 ],
       [-1.04008484, -2.4490238 , -0.1776889 , -0.29851096],
       [ 0.73110978, -0.

# in this step as we choose 2 components so we will have 2 vectors that carry the projections of other 2 features on them.so the output is only 2 coulumn.

In [0]:
from sklearn.decomposition import PCA

pca1 = PCA(n_components=2)
pcxtrain = pca1.fit_transform(X1_train)
pcxtest = pca1.transform(X1_test)
pcxtrain

array([[  1.27228206e+00,   3.57878727e-01],
       [  1.52231770e-01,  -2.99835283e-01],
       [ -2.18764183e+00,   6.17154256e-01],
       [  9.41913396e-01,   1.21810481e-02],
       [  1.76227837e+00,  -2.71066387e-01],
       [ -8.16806401e-02,  -1.04319475e+00],
       [ -1.89794863e+00,   1.48464927e+00],
       [  9.59395493e-01,  -6.91087802e-01],
       [  3.65661204e-01,  -9.13013087e-02],
       [  2.31845912e-01,  -1.01891237e+00],
       [  1.79730127e+00,   3.25092688e-01],
       [ -2.40350676e+00,  -2.44874760e-01],
       [  1.82199968e+00,   5.93295049e-01],
       [ -1.98427118e+00,   5.55515137e-01],
       [ -2.17073306e+00,   1.13527621e+00],
       [ -2.55598076e-01,  -2.67039928e+00],
       [  1.38862320e+00,   1.82768524e-01],
       [  1.93409268e+00,   9.57133382e-01],
       [  1.12243099e+00,  -1.78242941e+00],
       [  1.88169499e+00,  -8.17365689e-01],
       [  5.43123767e-02,  -1.32586239e+00],
       [  2.70365641e+00,   7.16489102e-01],
       [  

# For prediction we are using Randomforest classifier to predict 'y' that is label. you are eligible to use other methods for prediction. 

In [0]:

from sklearn.ensemble import RandomForestClassifier

classifier2 = RandomForestClassifier(max_depth=2, random_state=0)
classifier2.fit(pcxtrain, y1_train)

# Predicting the Test set results
y1_pred = classifier2.predict(pcxtest)
y1_pred

array(['Iris-virginica', 'Iris-versicolor', 'Iris-setosa',
       'Iris-virginica', 'Iris-setosa', 'Iris-virginica', 'Iris-setosa',
       'Iris-virginica', 'Iris-virginica', 'Iris-versicolor',
       'Iris-versicolor', 'Iris-virginica', 'Iris-versicolor',
       'Iris-versicolor', 'Iris-versicolor', 'Iris-setosa',
       'Iris-versicolor', 'Iris-versicolor', 'Iris-setosa', 'Iris-setosa',
       'Iris-versicolor', 'Iris-versicolor', 'Iris-setosa', 'Iris-setosa',
       'Iris-virginica', 'Iris-setosa', 'Iris-setosa', 'Iris-versicolor',
       'Iris-versicolor', 'Iris-setosa'], dtype=object)

In [0]:
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score


cm2 = confusion_matrix(y1_test, y1_pred)
print(cm2)
print('accuracy=',accuracy_score(y1_test, y1_pred))

[[11  0  0]
 [ 0 10  3]
 [ 0  2  4]]
accuracy= 0.833333333333


# Results with 3 components that means project other fetures in 3 vectors. so it would be 3 dimension.
Now let's try to evaluate classification performance of the random forest algorithm with 3 principal components. Update this piece of code:

In [0]:
import numpy as np
import pandas as pd
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"
names = ['sepal-length', 'sepal-width', 'petal-length', 'petal-width', 'Class']
dataset = pd.read_csv(url, names=names)
dataset.head()
X2 = dataset.drop('Class',1)
y2 = dataset['Class']

In [0]:
from sklearn.model_selection import train_test_split

X2_train, X2_test, y2_train, y2_test = train_test_split(X2, y2, test_size=0.2, random_state=0)


In [0]:
from sklearn.preprocessing import StandardScaler

sc2 = StandardScaler()
X2_train = sc2.fit_transform(X2_train)
X2_test = sc2.transform(X2_test)

In [0]:
from sklearn.decomposition import PCA

pca2 = PCA(n_components=3)
pcx2train = pca2.fit_transform(X2_train)
pcx2test = pca2.transform(X2_test)


In [0]:
from sklearn.ensemble import RandomForestClassifier

classifier3 = RandomForestClassifier(max_depth=2, random_state=0)
classifier3.fit(pcx2train, y2_train)

# Predicting the Test set results
y2_pred = classifier3.predict(pcx2test)


In [0]:
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score

cm3 = confusion_matrix(y2_test, y2_pred)
print(cm3)
print('accuracy=',accuracy_score(y2_test, y2_pred))

[[11  0  0]
 [ 0 12  1]
 [ 0  1  5]]
accuracy= 0.933333333333


as a summary, in this example with decision tree classifier for having 1component the accuracy is 0.933, 2component accuracy is 0.83 and 3component accuracy is 0.933. 

# Discussion
From the above experimentation we achieved optimal level of accuracy while significantly reducing the number of features in the dataset. We saw that accuracy achieved with only 1 principal component is equal to the accuracy achieved with will feature set i.e. 93.33%. It is also pertinent to mention that the accuracy of a classifier doesn't necessarily improve with increased number of principal components. From the results we can see that the accuracy achieved with one principal component (93.33%) was greater than the one achieved with two principal components (83.33%).

The number of principal components to retain in a feature set depends on several conditions such as storage capacity, training time, performance, etc. In some dataset all the features are contributing equally to the overall variance, therefore all the principal components are crucial to the predictions and none can be ignored. A general rule of thumb is to take number of principal of principal components that contribute to significant variance and ignore those with diminishing variance returns. A good way is to plot the variance against principal components and ignore the principal components with diminishing values as shown in the following graph: